win32ss/user/conime/conime.c

```diff
@@ -0,0 +1,1627 @@
+{
+    ULONG cbReturn;
+    LUID luid;
+    HANDLE hToken;
+    TOKEN_STATISTICS ts;
+
+    ZeroMemory(&luid, sizeof(luid));
+    if (NT_SUCCESS(NtOpenProcessToken(INVALID_HANDLE_VALUE, TOKEN_QUERY, &hToken)))
+    {
+        NtQueryInformationToken(hToken, TokenStatistics, &ts, sizeof(ts), &cbReturn);
+        NtClose(hToken);
+
+        RtlCopyLuid(&luid, &ts.AuthenticationId);
+
+        /* The LUID for the System account's logon session is always 999 */
+        if (!(luid.LowPart == 999 && luid.HighPart == 0))
+            g_bIsLogOnSession = FALSE;
+    }
+
+    return g_bIsLogOnSession;
+}
+
+/* @implemented */
+BOOL IntSendCopyDataTimeout(HWND hwndConsole, HWND hwndSender, PCOPYDATASTRUCT pcds)
+{
+    LRESULT ret;
+    DWORD_PTR result;
+
+    if (!hwndConsole)
+        return FALSE;
+
+    ret = SendMessageTimeoutW(hwndConsole, WM_COPYDATA, (WPARAM)hwndSender, (LPARAM)pcds,
+                              SMTO_ABORTIFHUNG, 3000, (PDWORD_PTR)&result);
+    return ret != 0;
+}
+
+PENTRY IntGetEnabledEntry(BOOL bEnabled)
+{
+    INT iEntry;
+    PENTRY pEntry;
+
+    RtlEnterCriticalSection(&g_csLock);
+    if (!g_bEnabled)
+        g_bEnabled = bEnabled;
+
+    if (g_cEntries <= 1)
+    {
+        RtlLeaveCriticalSection(&g_csLock);
+        return NULL;
+    }
+
+    iEntry = 1;
+    for (;;)
+    {
+        pEntry = g_pUserData->pEntries[iEntry];
+        if (pEntry)
+        {
+            if (pEntry->bEnabled == bEnabled && !pEntry->bWndEnabled)
+                break;
+        }
+
+        if (++iEntry >= g_cEntries)
+        {
+            RtlLeaveCriticalSection(&g_csLock);
+            return NULL;
+        }
+    }
+
+    pEntry = g_pUserData->pEntries[iEntry];
+    RtlLeaveCriticalSection(&g_csLock);
+    return pEntry;
+}
+
+BOOL IntGetLayoutText(PENTRY pEntry)
+{
+    HIMC hOldIMC = pEntry->hOldIMC;
+    HKL hKL = pEntry->hKL;
+    WCHAR szLayoutName[256];
+    WCHAR szKey[MAX_PATH];
+    HKEY phkResult;
+    DWORD cbData;
+
+    FIXME("\n");
+
+    pEntry->szLayoutText[0] = 0;
+
+    if (ImmEscapeW(hKL, hOldIMC, IME_ESC_IME_NAME, szLayoutName) ||
+        ImmGetIMEFileNameW(pEntry->hKL, szLayoutName, _countof(szLayoutName)))
+    {
+        StringCchCopyW(pEntry->szLayoutText, _countof(pEntry->szLayoutText), szLayoutName);
+    }
+    else
+    {
+        if (!GetKeyboardLayoutNameW(szLayoutName) || (szLayoutName[0] != 'E' && szLayoutName[0] != 'e'))
+            return FALSE;
+
+        StringCchCopyW(szKey, _countof(szKey), L"SYSTEM\\CurrentControlSet\\Control\\Keyboard Layouts\\");
+        StringCchCatW(szKey, _countof(szKey), szLayoutName);
+
+        if (RegOpenKeyExW(HKEY_LOCAL_MACHINE, szKey, 0, 9u, &phkResult))
+            return FALSE;
+
+        cbData = sizeof(pEntry->szLayoutText);
+        RegQueryValueExW(phkResult, L"layout text", 0, 0, (LPBYTE)pEntry->szLayoutText, &cbData);
+        RegCloseKey(phkResult);
+
+        if (pEntry->szLayoutText[0])
+        {
+            FIXME("\n");
```

Comment:
It would be more instructive to add an explanation about what should be fixed. (Here and at all the other places where such a useless "FIXME" is added.)